### ISSUE TITLE
Thoughts on tightening up the spec

### ISSUE THREAD

**Issue by lexaknyazev** [KhronosGroup/glTF#563]
### Base spec
- [x] 1. There is no integers in JS, but other languages can have them. Spec is silent on implicit type casts of integers provided for values declared as `FLOAT` or `FLOAT_VEC*`. I believe that a little permissive comment about that would be enough.
  - No spec update, https://github.com/KhronosGroup/glTF/issues/563#issuecomment-223148438
- [x] 2. `MODELVIEWINVERSETRANSPOSE` and `MODELINVERSETRANSPOSE` technique parameter semantics are declared as `FLOAT_MAT3`. At the same time presence of `node` field requires parameter to be `FLOAT_MAT4`. Spec is silent on this ambiguity. (see `boxSemantics/glTF/boxSemantics.gltf`)
  - Updated spec, f548e60e335a0ea3b7c039c5950dd823a83fd31b
- [x] 3. Is it OK to have technique parameters that are mentioned neither in `technique.uniforms` nor in `technique.attributes`?
  - Updated spec, ff4ac9797082bce00011e337f2790793f7a52b51
- [x] 4. Can technique parameter be referenced in `technique.attributes` and have a `value` property at the same time?
  - Updated spec, 537c356cc4023b86084b10a90b4d998908f6896a
- [x] 5. Can technique parameter with `FLOAT` (or other scalar) type and `"count": 3` be linked with `FLOAT_VEC3` accessor? (More details in the comment below.)
  - Updated spec, #614
- [x] 6. Are `{POSITION|NORMAL|...}_%d` semantics valid? Provided samples use only `TEXTURE_%d`.
  - Updated spec, b93fa35991ba58f8b1715a68f409ff749da0e2bc
- [x] 7. There is no explicit requirement on the `mesh.primitive.indices` accessor to not to be `FLOAT` and to reference `bufferView` with `ELEMENT_ARRAY_BUFFER` target.
  - Updated spec, #615
- [x] 8. Default material example (Appendix A) has an extra `a_texcoord_0` attribute.
  - Updated example, #608
- [x] 9. Also I think that spec could be more verbose on required semantics for mesh to be rendered with default material. Without `NORMAL` runtime can try to calculate them on-the-fly, but without  `POSITION` there's nothing to render at all.
  - Updated example, #608
- [x] 10. Do we need 2D versions of default material shaders for pure 2D content? Or in that case users must provide shader code themselves?
  - No, https://github.com/KhronosGroup/glTF/issues/563#issuecomment-223273802
- [x] 11. Spec [refers](https://github.com/KhronosGroup/glTF/tree/master/specification#specifying-extensions) to `EXT_binary_glTF` instead of `KHR_binary_glTF`.
  - Updated spec, #609 
- [x] 12. Spec lacks examples of using WebGL extensions, i.e., how to declare usage of `OES_element_index_uint`?
  `millennium_falcon.gltf` from `fr24-3d-models` has accessors with `"componentType": 5125` (`UNSIGNED_INT`) which isn't a valid value without extension.

Added from comments below
- [x] "Spec has a fixed list of valid attribute semantics. So how can we have custom attributes?", https://github.com/KhronosGroup/glTF/issues/563#issuecomment-211000702
  - Updated spec, 17d3145938c3086bdb9cca4b0f2d9cb9234889ee
- [ ] ~~"There is no way to specify texture's width and height. This is needed for using raw binary texture data (not web-compressed formats).", https://github.com/KhronosGroup/glTF/issues/563#issuecomment-212919091, https://github.com/KhronosGroup/glTF/issues/563#issuecomment-214246471~~
  - Moved to #620
- [x] Update `maximumAnisotropy` example, https://github.com/KhronosGroup/glTF/issues/563#issuecomment-218509248
  - Updated example, #613
- [x] Update `aspectRatio` default, https://github.com/KhronosGroup/glTF/issues/563#issuecomment-220009663
  - Updated spec, aa152d893d1db458698164e6254c820e79b27163
### KHR_binary_glTF
- [x] There is no requirement on `sceneSize` to be > 0.
  - Updated extension, #612 
### KHR_materials_common
- [x] Default value for `falloffAngle` is `PI/2` in spec but `PI` in `light.spot.schema.json`.
  - Updated extension, #611
### WEB3D_quantized_attributes
- [ ] ~~1. Should `decodedMax` and `decodedMin` array lengths be aligned with accessor's `type`?~~
- [ ] ~~2. Can we enforce `decodeMatrix` length to be just one of [4, 9 16] instead of [1-16]?~~

Moved to #610

---

_Original comment:_
### Base spec
1. There is no integers in JS, but other languages can have them. Spec is silent on implicit type casts of integers provided for values declared as `FLOAT` or `FLOAT_VEC*`. I believe that a little permissive comment about that would be enough.
2. `MODELVIEWINVERSETRANSPOSE` and `MODELINVERSETRANSPOSE` technique parameter semantics are declared as `FLOAT_MAT3`. At the same time presence of `node` field requires parameter to be `FLOAT_MAT4`. Spec is silent on this ambiguity. (see `boxSemantics/glTF/boxSemantics.gltf`)
3. Is it OK to have technique parameters that are mentioned neither in `technique.uniforms` nor in `technique.attributes`?
4. Can technique parameter be referenced in `technique.attributes` and have a `value` property at the same time?
5. ~~Can technique parameter with `FLOAT` (or other scalar) type and `"count": 3` be linked with `FLOAT_VEC3` accessor?~~ More details in the comment below.
6. Are `{POSITION|NORMAL|...}_%d` semantics valid? Provided samples use only `TEXTURE_%d`.
7. There is no explicit requirement on the `mesh.primitive.indices` accessor to not to be `FLOAT` and to reference `bufferView` with `ELEMENT_ARRAY_BUFFER` target.
8. Default material example (Appendix A) has an extra `a_texcoord_0` attribute.
9. Also I think that spec could be more verbose on required semantics for mesh to be rendered with default material. Without `NORMAL` runtime can try to calculate them on-the-fly, but without  `POSITION` there's nothing to render at all.
10. Do we need 2D versions of default material shaders for pure 2D content? Or in that case users must provide shader code themselves?
11. Spec [refers](https://github.com/KhronosGroup/glTF/tree/master/specification#specifying-extensions) to `EXT_binary_glTF` instead of `KHR_binary_glTF`.
12. Spec lacks examples of using WebGL extensions, i.e., how to declare usage of `OES_element_index_uint`?
    `millennium_falcon.gltf` from `fr24-3d-models` has accessors with `"componentType": 5125` (`UNSIGNED_INT`) which isn't a valid value without extension.
### KHR_binary_glTF

There is no requirement on `sceneSize` to be > 0.
### KHR_materials_common

Default value for `falloffAngle` is `PI/2` in spec but `PI` in `light.spot.schema.json`.
### WEB3D_quantized_attributes
1. Should `decodedMax` and `decodedMin` array lengths be aligned with accessor's `type`?
2. Can we enforce `decodeMatrix` length to be just one of [4, 9 16] instead of [1-16]?

**Comment by pjcozzi** [comment]
Thanks @lexaknyazev!  This is great feedback.  I created a `1.0.1` label and tagged this.  I'll do some spec updates for the obvious ones and get back to you with questions on any others.  I have a few things on my plate right now so please give me some time.

**Comment by lexaknyazev** [comment]
Let's say we have an `accessor` with `"type": "VEC3"` and `"componentType": 5126 //FLOAT`.
Are both kinds of technique parameters (see below) allowed to reference it?
```js
 "technique0": {
      "attributes": {
        "a_1": "param1",
        "a_2": "param2"
      },
      "parameters": {
        "param1": {
          "type": 35665 // FLOAT_VEC3
        },
       "param2": {
          "count": 3,
          "type": 5126 // FLOAT
        },
      }
    }
```

**Comment by lexaknyazev** [comment]
Spec has a fixed list of valid attribute semantics. So how can we have custom attributes?
There was a [comment](https://github.com/KhronosGroup/glTF/issues/239#issuecomment-36097457) about that but spec doesn't mention anything like that.

**Comment by pjcozzi** [comment]
> Let's say we have an accessor with "type": "VEC3" and "componentType": 5126 //FLOAT.
Are both kinds of technique parameters (see below) allowed to reference it?

I don't think the spec defines this yet (or at least not well), but perhaps only `param1` is valid.  Do you have a use case for allowing `param2`?

> Spec has a fixed list of valid attribute semantics. So how can we have custom attributes?

Semantics are optional: https://github.com/KhronosGroup/glTF/tree/master/specification#techniqueparameters-1

So a custom attribute would need to rely on a well-known attribute name - not semantic - to conform to the spec as it is written now.  I think the question should be: can we have custom semantics, right?  I think the answer should be yes, but we need to think about how to do it in a forward-compatible way, and how important that would be in practice (e.g., if an asset uses a custom semantic and then a later version of glTF defines that semantic, but it is different than the custom one's meaning; given that the future glTF version could have other breaking changes, it might be OK).

**Comment by lexaknyazev** [comment]
Please correct me if I'm wrong, we have:
`mesh.primitive.attributes: {semantic: accessor_id, ...}`
`technique.attributes: {attributeName: parameterName, ...}`
`technique.parameters: {parameterName: {"semantic": semantic}, ...}`
If attribute technique parameter doesn't have a semantic how can it be linked to an accessor?

**Comment by lexaknyazev** [comment]
There is no way to specify texture's width and height. This is needed for using raw binary texture data (not web-compressed formats).

**Comment by lexaknyazev** [comment]
>  I think the question should be: can we have custom semantics, right? I think the answer should be yes, but we need to think about how to do it in a forward-compatible way

Imho, the simplest way to achieve forward-compatibility is to require all custom semantics to have some prefix (such as `_`) so they never collide with spec-defined names.

**Comment by lexaknyazev** [comment]
Looking at the [extensions.json example](https://github.com/KhronosGroup/glTF/blob/master/specification/schema/examples/extensions.json) I think we need to have some standardized way to define external extensions schemes. I see at least three possible approaches here:

1. Explicitly define extension-specific schemes for all possible extensions like

    ```js
{
        "$schema" : "http://json-schema.org/draft-03/schema",
        "title" : "EXT_texture_filter_anisotropic",
        "type" : "object",
        "properties" : {
            "maximumAnisotropy" : {
                "type" : "integer",
                "default" : "0"
            }
        },
        "additionalProperties" : false
}
```
2. Do not define any additional schemes and just use appropriate GL names (`TEXTURE_MAX_ANISOTROPY_EXT` in this case) instead of js-like camelCase.

    ```js
    {
        "extensionsUsed" : ["EXT_texture_filter_anisotropic"],
        "samplers" : {
            "sampler_id": {
                "magFilter": 9729,
                "minFilter": 9987,
                "name": "user-defined sampler name",
                "wrapS": 10497,
                "wrapT": 10497,
                "extensions" : {
                    "EXT_texture_filter_anisotropic" : {
                        "TEXTURE_MAX_ANISOTROPY_EXT" : 8
                    }
                }
            }
        }
    }
```
3. Do not use text-based properties and reference parameters as corresponding GL consts.

    ```js
    {
        "extensionsUsed" : ["EXT_texture_filter_anisotropic"],
        "samplers" : {
            "sampler_id": {
                "magFilter": 9729,
                "minFilter": 9987,
                "name": "user-defined sampler name",
                "wrapS": 10497,
                "wrapT": 10497,
                "extensions" : {
                    "EXT_texture_filter_anisotropic" : {
                        "parameters": [
                            {
                                "pname": 0x84FE, //TEXTURE_MAX_ANISOTROPY_EXT
                                "value": 8
                            }
                        ]
                    }
                }
            }
        }
    }
```

And for extensions like `OES_element_index_uint` (with no additional properties or functions) maybe it should be sufficient just mentioning them in `extensionsUsed` property of the root object.

**Comment by lexaknyazev** [comment]
> There is no way to specify texture's width and height. This is needed for using raw binary texture data (not web-compressed formats).

After more digging, I propose to add optional `width`, `height` and `mimeType` fields to `image` and remove them from `KHR_binary_glTF\image`. This will serve two purposes:

1. We'll be able to use uncompressed textures both from external files (with explicit `width`, `height` and `"mimeType": "application/octet-stream"`) and from embedded `data:` URI. In the latter case, we should omit `mimeType` field and use URI: `data:application/octet-stream;base64,...`.

2. `KHR_binary_glTF\image` will contain only `bufferView` just like `KHR_binary_glTF\shader`.

**Comment by pjcozzi** [comment]
> Please correct me if I'm wrong, we have:
mesh.primitive.attributes: {semantic: accessor_id, ...}
technique.attributes: {attributeName: parameterName, ...}
technique.parameters: {parameterName: {"semantic": semantic}, ...}
If attribute technique parameter doesn't have a semantic how can it be linked to an accessor?

Looking at this again, I agree there is a disconnect here that we need to address in the spec.  Will need to brainstorm if it is as simple as requiring the semantic.

>>I think the question should be: can we have custom semantics, right? I think the answer should be yes, but we need to think about how to do it in a forward-compatible way
> Imho, the simpliest way to achieve forward-compatibility is to require all custom semantics to have some prefix (such as _) so they never collide with spec-defined names.

That could work.

> Looking at the extensions.json example I think we need to have some standardized way to define external extensions schemes. I see at least three possible approaches here:

Each extension defines the JSON schema for the extra JSON properties, e.g., see https://github.com/KhronosGroup/glTF/blob/master/extensions/Vendor/CESIUM_RTC/CESIUM_RTC.schema.json from https://github.com/KhronosGroup/glTF/blob/master/extensions/Vendor/CESIUM_RTC/README.md

What else are you looking for here?

> And for extensions like OES_element_index_uint (with no additional properties or functions) maybe it should be sufficient just mentioning them in extensionsUsed property of the root object.

This would be a bit of a magic value; I still think we need the extension explicitly defined even if it is quite short and simple.

>> There is no way to specify texture's width and height. This is needed for using raw binary texture data (not web-compressed formats).
>After more digging, I propose to add optional width, height and mimeType fields to image and remove them from KHR_binary_glTF\image. This will serve two purposes:

>We'll be able to use uncompressed textures both from external files (with explicit width, height and "mimeType": "application/octet-stream") and from embedded data: URI. In the latter case, we should omit mimeType field and use URI: data:application/octet-stream;base64,....

>KHR_binary_glTF\image will contain only bufferView just like KHR_binary_glTF\shader.

This could work well.  We should also consider it in the context of texture compression so we don't end of changing it again for that (additions would be fine, of course).

**Comment by lexaknyazev** [comment]
> Each extension defines the JSON schema for the extra JSON properties, e.g., see https://github.com/KhronosGroup/glTF/blob/master/extensions/Vendor/CESIUM_RTC/CESIUM_RTC.schema.json from https://github.com/KhronosGroup/glTF/blob/master/extensions/Vendor/CESIUM_RTC/README.md
What else are you looking for here?

Some extensions (like `OES_texture_float` or `OES_standard_derivatives`) are already well-defined up to 3 times (OpenGL, OpenGL ES, WebGL). They don't have anything glTF-specific.

Do we _really_ need glTF schemes for them? The lack of some sort of "fast lane" could slow down development of  glTF implementations.

**Comment by pjcozzi** [comment]
> Some extensions (like OES_texture_float or OES_standard_derivatives) are already well-defined up to 3 times (OpenGL, OpenGL ES, WebGL). They don't have anything glTF-specific.

> Do we really need glTF schemes for them?

Sorry for the slow response; I was at a conference all last week.

We do need an extension (.md file) for them, but it can be very small and just reference, for example, the GL extension.  In some cases, the extension will not require any new schema, in other cases, it may just be new allowed enums, and in other cases, it may be new properties/objects.

**Comment by lexaknyazev** [comment]
> We do need an extension (.md file) for them

Of course.

> it may be new properties/objects

My point is that we need conventions for translation GL names into new properties, e.g., how `TEXTURE_MAX_ANISOTROPY_EXT` become `maximumAnisotropy`?

**Comment by pjcozzi** [comment]
> My point is that we need conventions for translation GL names into new properties, e.g., how TEXTURE_MAX_ANISOTROPY_EXT become maximumAnisotropy?

Ah, yes, I agree throughout glTF, not just extensions.  `maximumAnisotropy` is just in an example, not in an actual extension, but it should copy the GL name, e.g., `maxAnisotropy` or `textureMaxAnisotropy`.

**Comment by lexaknyazev** [comment]
Does `[camera.perspective.]aspectRatio: 0` mean the same as undefined `aspectRatio`?

**Comment by pjcozzi** [comment]
> Does [camera.perspective.]aspectRatio: 0 mean the same as undefined aspectRatio?

The spec is not explicit about this.  Perhaps the 1.0.1 spec should change this from `>= 0` to`> 0`?

**Comment by lexaknyazev** [comment]
> Perhaps the 1.0.1 spec should change this from `>= 0` to `> 0`?

Makes sense.

**Comment by pjcozzi** [comment]
> There is no integers in JS, but other languages can have them. Spec is silent on implicit type casts of integers provided for values declared as FLOAT or FLOAT_VEC*. I believe that a little permissive comment about that would be enough.

@lexaknyazev do you have a specific example for this?  JSON schema actually allows us to define fields as integers, e.g., [accessor. byteOffset](https://github.com/KhronosGroup/glTF/tree/master/specification#accessor).

**Comment by lexaknyazev** [comment]
`shininess` is defined as `FLOAT`
https://github.com/KhronosGroup/glTF/blob/master/sampleModels/Box/glTF/Box.gltf#L226
value is integer
https://github.com/KhronosGroup/glTF/blob/master/sampleModels/Box/glTF/Box.gltf#L90

**Comment by pjcozzi** [comment]
>  2. MODELVIEWINVERSETRANSPOSE and MODELINVERSETRANSPOSE technique parameter semantics are declared as FLOAT_MAT3. At the same time presence of node field requires parameter to be FLOAT_MAT4. Spec is silent on this ambiguity. (see boxSemantics/glTF/boxSemantics.gltf)

@lexaknyazev currently, this implies that a parameter with `MODELVIEWINVERSETRANSPOSE` or `MODELINVERSETRANSPOSE` semantic can't have a `node` property so it, therefore, can't get it's value from a node's transform.  I think this is fine.  I can make it more explicit if you want.  Did you have other thoughts?

**Comment by lexaknyazev** [comment]
So it looks like there was a bug in collada2gltf some time ago:
https://github.com/KhronosGroup/glTF/blob/d453cd16420d4f3c8a0bf4a1f1c7a9c8c4f70fbc/sampleModels/boxSemantics/glTF/boxSemantics.gltf#L407

I think spec can be explicit on that to prevent runtimes from somehow "converting" matrices.
Strictly speaking, `MODELINVERSETRANSPOSE`, `MODELVIEWINVERSETRANSPOSE` and `VIEWPORT` semantics can't coexist with `node` property.

**Comment by pjcozzi** [comment]
> I think spec can be explicit on that to prevent runtimes from somehow "converting" matrices.
Strictly speaking, MODELINVERSETRANSPOSE, MODELVIEWINVERSETRANSPOSE and VIEWPORT semantics can't coexist with node property.

Thanks, updated the spec in f548e60e335a0ea3b7c039c5950dd823a83fd31b.  I will also cherry-pick this for master since it does not depend on the 1.0.1 spec.

BTW, pull requests into the [1.0.1 branch](https://github.com/KhronosGroup/glTF/compare/1.0.1) for any issues are welcome; in the meantime, I am working through items in this issue in the next few days.

**Comment by pjcozzi** [comment]
> shininess is defined as FLOAT
https://github.com/KhronosGroup/glTF/blob/master/sampleModels/Box/glTF/Box.gltf#L226
value is integer
https://github.com/KhronosGroup/glTF/blob/master/sampleModels/Box/glTF/Box.gltf#L90

I suspect this is because `JSON.stringify` or equivalent just dropped the `.0`.  JSON just stores `Number` and it is OK for a floating-point value to be a whole number.  A clarification here just seems unnecessary.  Are you sure it would be useful?

**Comment by lexaknyazev** [comment]
> JSON just stores Number and it is OK for a floating-point value to be a whole number. A clarification here just seems unnecessary. 

OK. 

(If only GLSL was so permissive...)

**Comment by pjcozzi** [comment]
> 3. Is it OK to have technique parameters that are mentioned neither in `technique.uniforms` nor in `technique.attributes`?

No, but the spec wasn't explicit.  Updated spec in the 1.0.1 branch: ff4ac9797082bce00011e337f2790793f7a52b51

**Comment by pjcozzi** [comment]
>> JSON just stores Number and it is OK for a floating-point value to be a whole number. A clarification here just seems unnecessary.
> OK.

> (If only GLSL was so permissive...)

lol.  Thanks for the quick response.  It is much appreciated and makes updating the spec fast!

**Comment by pjcozzi** [comment]
> 4. Can technique parameter be referenced in `technique.attributes` and have a `value` property at the same time?

No, updated the spec, 537c356cc4023b86084b10a90b4d998908f6896a.

This would create an implementation burden for a rarely needed feature.

**Comment by pjcozzi** [comment]
> Are `{POSITION|NORMAL|...}_%d` semantics valid? Provided samples use only `TEXTURE_%d`.

Updated spec, b93fa35991ba58f8b1715a68f409ff749da0e2bc, to be explicit that this is only valid for texture coordinates and colors.  I am also open to extending this up to other attribute semantics if you have specific examples where the added complexity would be useful in practice.

**Comment by pjcozzi** [comment]
> 8. Default material example (Appendix A) has an extra a_texcoord_0 attribute.
> 9. Also I think that spec could be more verbose on required semantics for mesh to be rendered with default material. Without NORMAL runtime can try to calculate them on-the-fly, but without POSITION there's nothing to render at all.

Update in #608.

I removed the `NORMAL` attribute since it is not needed for this simple example.

**Comment by pjcozzi** [comment]
>  10. Do we need 2D versions of default material shaders for pure 2D content? Or in that case users must provide shader code themselves?

Probably overkill for now.  If there is enough community interest, we could add an example.

**Comment by pjcozzi** [comment]
> 11. Spec refers to EXT_binary_glTF instead of KHR_binary_glTF.

Good eye.  Updated in #609.

**Comment by lexaknyazev** [comment]
>  I am also open to extending this up to other attribute semantics if you have specific examples where the added complexity would be useful in practice.

Maybe we should endorse [custom semantics](https://github.com/KhronosGroup/glTF/issues/563#issuecomment-213397568) instead?

**Comment by pjcozzi** [comment]
>> I am also open to extending this up to other attribute semantics if you have specific examples where the added complexity would be useful in practice.
> Maybe we should endorse custom semantics instead?

I was thinking the same thing - just haven't gotten that far in this issue yet.  Will follow-up soon.

**Comment by pjcozzi** [comment]
> Default value for falloffAngle is PI/2 in spec but PI in light.spot.schema.json.

See #611

**Comment by pjcozzi** [comment]
> KHR_binary_glTF - There is no requirement on sceneSize to be > 0.

See #612

**Comment by pjcozzi** [comment]
>> Does [camera.perspective.]aspectRatio: 0 mean the same as undefined aspectRatio?
> The spec is not explicit about this. Perhaps the 1.0.1 spec should change this from >= 0 to> 0?

Looking at this, https://github.com/KhronosGroup/glTF/issues/563#issuecomment-220009663, again, I think it is OK as is.  An aspect ratio of zero means that the y field of view is also zero (not the same as the `aspectRatio` property being `undefined`), and these values are commonly allowed in perspective matrix functions, for example, https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Core/Matrix4.js#L660-L667.

What do you think?

**Comment by lexaknyazev** [comment]
> What do you think?

[L663](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Core/Matrix4.js#L663) throws on `fovY == 0`, while error says that zero is allowed. It's vice-versa for `PI`.

Zero `fovY` will cause division by zero on [L681](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Core/Matrix4.js#L681).

Moreover, [L665](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Core/Matrix4.js#L665) demands `aspectRatio` to be greater than zero.

**Comment by pjcozzi** [comment]
> L663 throws on fovY == 0, while error says that zero is allowed. It's vice-versa for PI.

> Zero fovY will cause division by zero on L681.

> Moreover, L665 demands aspectRatio to be greater than zero.

Ah, yes, I had that backwards.

Will update so fovY and aspect ratio must be `> 0`, not `>= 0`.

**Comment by pjcozzi** [comment]
> Will update so fovY and aspect ratio must be > 0, not >= 0.

See aa152d893d1db458698164e6254c820e79b27163.

**Comment by pjcozzi** [comment]
> "Spec has a fixed list of valid attribute semantics. So how can we have custom attributes?", https://github.com/KhronosGroup/glTF/issues/563#issuecomment-211000702

See 17d3145938c3086bdb9cca4b0f2d9cb9234889ee.

**Comment by pjcozzi** [comment]
>> There is no way to specify texture's width and height. This is needed for using raw binary texture data (not web-compressed formats).

> After more digging, I propose to add optional width, height and mimeType fields to image and remove them from KHR_binary_glTF\image. This will serve two purposes:

> We'll be able to use uncompressed textures both from external files (with explicit width, height and "mimeType": "application/octet-stream") and from embedded data: URI. In the latter case, we should omit mimeType field and use URI: data:application/octet-stream;base64,....

> KHR_binary_glTF\image will contain only bufferView just like KHR_binary_glTF\shader.

After evaluating this, https://github.com/KhronosGroup/glTF/issues/563#issuecomment-214246471, more, I agree that this is the right direction; however, it will require a new version of the KHR_binary_glTF extension to take full advantage.

I'd like to limit the scope of the 1.0.1 spec largely to improvements for validation (a few exceptions are OK, for example, #573).

I feel like we might be better off holding off on this until glTF includes texture compression and then making all the changes at once: this, texture compression, and KHR_binary_glTF updates.

What do you think?  Do you have pressing use cases that should drive this happening sooner?

**Comment by lexaknyazev** [comment]
> What do you think?  Do you have pressing use cases that should drive this happening sooner?

Several points, actually:

1. Uncompressed is the simplest possible pixel format. All GL implementations support it. It looks like `texture.type` has no use without `KHR_binary_glTF` while spec & extension are in their current state.

2. We need to consider multiple versions of the same texture for different compression formats (it's up to runtime to select one from RAW/PVRTC/ETC/S3TC/JPG). Export tools can provide only uncompressed file while all others can be generated during pipeline processing.

3. Adding fields to `image` is both ways compatible, removing fields from `KHR_binary_glTF\image` is not. We have to ask `KHR_binary_glTF` implementations for feedback on this.

**Comment by lexaknyazev** [comment]
> I feel like we might be better off holding off on this until glTF includes texture compression and then making all the changes at once: this, texture compression, and KHR_binary_glTF updates.

Maybe we can establish some sort of framework for texture formats now to be able to add exact compression extensions later in non-destructive manner?

**Comment by lexaknyazev** [comment]
And we shouldn't forget about `FLOAT` textures - imho, the most important real-world uncompressed use case.

**Comment by pjcozzi** [comment]
@lexaknyazev are you up for proposing all these texture changes in a new issue (can still be labeled `1.0.1`)?  If we can nail this down enough - establish the framework as you suggest - than I think we can do it in the near-term.  As for `KHR_binary_glTF`, I'm afraid that we would have to create a new `KHR_binary_glTF2` (or something) to take advantage of this.  That could be postponed until the extension potential calls for other changes.

Also - one more use case for this: creating glTF materials procedurally at runtime.

**Comment by lexaknyazev** [comment]
> @lexaknyazev are you up for proposing all these texture changes in a new issue (can still be labeled 1.0.1)? 

OK

> As for KHR_binary_glTF, I'm afraid that we would have to create a new KHR_binary_glTF2 (or something) to take advantage of this. That could be postponed until the extension potential calls for other changes.

I understand that extension has been ratified already.
However it looks like almost all extension spec and [known implementations](https://github.com/KhronosGroup/glTF/tree/master/extensions/Khronos/KHR_binary_glTF#known-implementations) contributors can be reached for feedback on this particular change.

**Comment by pjcozzi** [comment]
>> As for KHR_binary_glTF, I'm afraid that we would have to create a new KHR_binary_glTF2 (or something) to take advantage of this. That could be postponed until the extension potential calls for other changes.

> I understand that extension has been ratified already.
However it looks like almost all extension spec and known implementations contributors can be reached for feedback on this particular change.

This extension is widely used in Cesium; it is the default.  We could keep backwards-compatible code in the loader, but new assets won't work with older versions of Cesium, which would mean the converter would need another check box (`[ ] Make compatible with Cesium 1.x and earlier`).

I would rather queue this extension spec change up with other potential changes for a v2.

**Comment by pjcozzi** [comment]
> 12. Spec lacks examples of using WebGL extensions, i.e., how to declare usage of OES_element_index_uint? millennium_falcon.gltf from fr24-3d-models has accessors with "componentType": 5125 (UNSIGNED_INT) which isn't a valid value without extension.

I originally said that this should be a glTF extension, https://github.com/KhronosGroup/glTF/issues/563#issuecomment-218497505.  I started writing the extension spec, and I now agree that it is overkill.

For the 1.0.1 spec, I suggest that we just allow `UNSIGNED_INT` as a valid `accessor.componentType` enum when the accessor is used for indices, not attributes (since this is not a valid value for [vertexAttribPointer](https://www.khronos.org/opengles/sdk/docs/man/xhtml/glVertexAttribPointer.xml)).

This means that the client will need to enable the OES_element_index_uint WebGL extension, or divide the accessor up (and all that entails) if the extension is not supported.  However, given that the extension is [widely supported](http://webglstats.com/), I do not think we need to worry about the later in practice.

What do you think?

**Comment by lexaknyazev** [comment]
I think this
> This means that the client will need to enable the OES_element_index_uint WebGL extension

is a crucial point. Client has to know about uint indices, so I'd prefer to mention `OES_element_index_uint` in the `extensionsUsed` array for WebGL 1.0 profile.

**Comment by pjcozzi** [comment]
> Client has to know about uint indices, so I'd prefer to mention OES_element_index_uint in the extensionsUsed array for WebGL 1.0 profile.

Yes, I thought about this too.  `extensionsUsed` is for _glTF_ extensions, not WebGL extensions so the right approach might be to introduce a new `webglExtensionsUsed` property so a client could check them and simply loop over and call `getExtension` if need be.

This could be convenient for clients and avoids double-tasking `extensionsUsed`.  It also maps well to other WebGL extensions like `OES_standard_derivatives` and new texture and attribute formats.

Thoughts?

**Comment by lexaknyazev** [comment]
+1 for separate list.

The situation can become a bit more complex for GLSL extensions like `EXT_shader_texture_lod`.
Shader code may compile with or without extension support because of preprocessor directives.
So runtime has to know somehow is rendering possible without such extension.

**Comment by lexaknyazev** [comment]
> This extension is widely used in Cesium; it is the default. We could keep backwards-compatible code in the loader, but new assets won't work with older versions of Cesium, which would mean the converter would need another check box ([ ] Make compatible with Cesium 1.x and earlier).

> I would rather queue this extension spec change up with other potential changes for a v2.

I believe it's possible for `1.0.1` to support raw textures and prepare some foundation for compression formats.
Changes to `KHR_binary_glTF` could be queued up for the next version to preserve compatibility.

**Comment by pjcozzi** [comment]
>> This extension is widely used in Cesium; it is the default. We could keep backwards-compatible code in the loader, but new assets won't work with older versions of Cesium, which would mean the converter would need another check box ([ ] Make compatible with Cesium 1.x and earlier).

> I would rather queue this extension spec change up with other potential changes for a v2.
I believe it's possible for 1.0.1 to support raw textures and prepare some foundation for compression formats.
Changes to KHR_binary_glTF could be queued up for the next version to preserve compatibility.

Yes, exactly.  To keep things organized, can you please start the proposal in a new issue?

**Comment by pjcozzi** [comment]
>  12. Spec lacks examples of using WebGL extensions, i.e., how to declare usage of OES_element_index_uint? millennium_falcon.gltf from fr24-3d-models has accessors with "componentType": 5125 (UNSIGNED_INT) which isn't a valid value without extension.

See #619.

**Comment by lexaknyazev** [comment]
> Yes, exactly. To keep things organized, can you please start the proposal in a new issue?
#620

**Comment by pjcozzi** [comment]
@lexaknyazev have all tasks in this issue been addressed or moved to other issues?  If so, I will label this as `resolved` (and then close once the 1.0.1 spec is merged).

**Comment by lexaknyazev** [comment]
I've found only one unaddressed issue in this thread: attribute parameters must have semantic, otherwise they will be unlinkable to accessors [(comment)](https://github.com/KhronosGroup/glTF/issues/563#issuecomment-214469715).

**Comment by pjcozzi** [comment]
> I've found only one unaddressed issue in this thread: attribute parameters must have semantic, otherwise they will be unlinkable to accessors [(comment)](https://github.com/KhronosGroup/glTF/issues/563#issuecomment-214469715).

Good eye.  What do you think of 28e113d277002404803b7618d86c1fc7caf71dd8?

**Comment by lexaknyazev** [comment]
> What do you think of 28e113d?

Lgtm

**Comment by pjcozzi** [comment]
Labeling this as resolved.  Please submit future 1.0.1 spec issues in new issues.

**Comment by pjcozzi** [comment]
Updated in #826
